src/sdk/strategy.ts

```diff
@@ -0,0 +1,121 @@
+import { SecretKey } from '@nucypher/nucypher-core';
+import { ethers } from 'ethers';
+
+import { EnactedPolicy } from '../policies/policy';
+import { Alice } from '../characters/alice';
+import { Bob } from '../characters/bob';
+import { Enrico } from '../characters/enrico';
+import { tDecDecrypter } from '../characters/universal-bob';
+
+import { Cohort } from './cohort';
+
+export class Strategy {
+  public policy?: EnactedPolicy;
+  public encrypter?: Enrico;
+  public decrypter?: tDecDecrypter;
+  constructor(
+    public cohort: Cohort,
+    public startDate: Date,
+    public endDate: Date,
+    private alice: Alice,
+    private bob: Bob,
+    private bobSecretKey: SecretKey,
+    public deployed: boolean
```

Comment:
I'd love to avoid having flags like this, and instead, return a new type `DeployedStrategy` that would expose relevant methods.

I've tried to lobby for this design in my SDK proposal version. WDYT? Is it overdone? Or maybe too complicated at this stage of work on SDK?

Reply:
I'm not sure what's best. I've pinged some people more wise than myself to see if they have strong opinions on this.

How would it relate to `revoking` a strategy? Is a revoked strategy no longer deployed? is there a more general case here?....not sure, thinking out loud tbh

Reply:
I think a revoked strategy is still technically deployed but it's no longer in effect - strategy users may not perform decryption anymore.

The way I imagined structuring the `Strategy` verbs:
```
Strategy.deploy => DeployedStrategy
DeployedStrategy.revoke => Strategy
Strategy.update => Strategy
``

Reply:
Given what you've said, `DeployedStrategy.revoke => DeployStrategy` no?

Can a revoked strategy be un-revoked?

Reply:
> Given what you've said, DeployedStrategy.revoke => DeployStrategy no?

I'd say it's the same as `Strategy` - it's not active, and it has to be deployed before it takes an effect.

> Can a revoked strategy be un-revoked? or is it basically dead

I think once a strategy is revoked, it should be treated as "dead". And if we'd like to "bring it back" we need to redeploy it etc.

It may be a good idea to avoid using `update` verb all-together, and instead rely on `revoke` and `deploy` to get the job done. Is there any practical reason to avoid that? I could see some merit in "atomic" updates (`update` over `revoke` & `deploy` combo) since it doesn't introduce any downtime.

Reply:
@KPrasch @jMyles @derekpierre @vepkenez @fjarri Any thoughts here on design best practices?

We'd like to maximise for:
- future proofing
- ease of understanding (by adopters)
- ease of testing within the library
- ease of testing in adopter code

Reply:
> It may be a good idea to avoid using the update verb altogether and instead rely on revoke and deploy to get the job done. Is there any practical reason to avoid that? I could see some merit in "atomic" updates (update over revoke & deploy combo) since it doesn't introduce any downtime.

Completely agree, I have actually already removed the `update` verb. `Strategy` is immutable, once it's been created it can only be deployed (maybe that's a good +1 in favour of not having `deployed` as an attribute...)

Reply:
> I'd love to avoid having flags like this, and instead, return a new type DeployedStrategy that would expose relevant methods.

We could completely flip this....

We have a `Deployer` class which has a `strategies` attribute of immutable `Strategy` objects (dict mapping label to object).
The deployer class can handle all of the `deploy`, `revoke`, `delete` verbs

Reply:
After reading the discussion, my initial thoughts here are:

If feels like the philosophy of the `Strategy` class is analogous to an immutable configuration, and not necessarily an object that does any execution.

As a result, I like the separate `Deployer` idea, as the thing that performs an action based on the Strategy. It also feels like encrypter/decrypter should not be members variables stored within a Strategy object.

So perhaps something like:
```
StrategyDeployer.deploy(label, strategy) -> Policy
...
StrategyDeployer.revoke(policy, strategy)
```

For the encrypter/decrypter, Strategy can then be
```
Stategy. getEncrypter(policy) -> Encrypter   # this call can check whether the policy is valid before returning

Strategy.getDecrypter(policy) -> Decrypter  # this call can check whether the policy is valid before returning
```

In this way Strategy never stores `policy`, `encrypter`, `decrypter` and is much more reusable across different policies. For example, the same Strategy can be used with different labels. 

wdyt?

Reply:
Yes, i like the idea of removing `encrypter` and `decrypter` from `Strategy`. I think those functions would also be moved to the Deployer

```
Deployer.getEncrypter(policy) -> Encrypter 
Deployer.getDecrypter(policy) -> Decrypter 
```